chore(service): Improvements for waiting on service readiness

rhuss opened the pull request:
* Increased default timeout to 600s. This timeout will be triggered
  only when the Ready condition stays in UNKNOWN for that long. If its
  True or False then the command will return anyway sooner.
  So it makes sense to go for a much longer timeout than 60s
* Enhanced output to indicate the progress

This change needs some updates to the API and introduces a 'MessageCallback'
type which is calle for each intermediate event with the "Ready" condition message.

Example of the UX:

[![asciicast](https://asciinema.org/a/5MwD9j3HJpLdBgosJFbFsDBdB.svg)](https://asciinema.org/a/5MwD9j3HJpLdBgosJFbFsDBdB)

maximilien commented on docs/cmd/kn_service_create.md:
```diff
@@ -59,7 +59,7 @@ kn service create NAME --image IMAGE [flags]
       --revision-name string     The revision name to set. Must start with the service name and a dash as a prefix. Empty revision name will result in the server generating a name for the revision. Accepts golang templates, allowing {{.Service}} for the service name, {{.Generation}} for the generation, and {{.Random [n]}} for n random consonants. (default "{{.Service}}-{{.Random 5}}-{{.Generation ...
       --service-account string   Service account name to set. Empty service account name will result to clear the service account.
-      --wait-timeout int         Seconds to wait before giving up on waiting for service to be ready. (default 60)
+      --wait-timeout int         Seconds to wait before giving up on waiting for service to be ready. (default 600)
```
Does it make sense to also have a `--no-wait`?

maximilien replied:
600 seconds == 10 minutes. Is that too much for a default? I feel like one minute (60 seconds) should be good enough. Thoughts?

rhuss replied:
We had 60s before, but that was often too short, especially when the created Pods still have to pull the image, which is often longer than one minute. In that case a "timeout error" occurs as in #423, which is confusing as the service itself is ready eventually.

As I'm assuming that as long the "Ready" condition is in the state "UNKNOWN" I rely on Knative that this eventually will be reconciled to either "TRUE" or "FALSE". If this is the case, the command will return. It will only run into a timeout when staying in UNKNOWN. So from my experience, we should rely on Knative serving to do that reconciliation and e.g. if it can't get the service up in a certain amount of time, will be put the "Ready" condition to "FALSE". Hence this large timeout, which could maybe even longer (not sure if there are image that large that it will take longer than 10 mins, but then a sync wait on the CLI without a real progress meter is tedious, too).

I added the detailed feedback on the event as a poor-men progress meter replacement, so that the user sees that something is going on even when the creation/update takes a bit longer.

That output is also helpful for scripting use cases as it helps in debugging timing issues.

Still, with a `-q` we should certainly not show that messages. 

One can argue that one migt move these event to `--verbose`, but then one would lose the simple progress indicator.

rhuss replied:
`--no-wait` is achieved with `--async` (that was chosen over no-wait at the time when the feature was introduced)

maximilien replied:
OK I can live with that. Thanks.

maximilien commented on pkg/wait/wait_for_ready.go:
```diff
@@ -38,8 +39,8 @@ type waitForReadyConfig struct {
 	// Wait on resource the resource with this name until a given timeout
-	// and write status out on writer
-	Wait(name string, timeout time.Duration) error
+	// and write event messages for uknown event to the status writer
```
*unknown

maximilien commented:
/ok-to-test

navidshaikh commented:
From e2e, `ContainsAll` comparisons need update.

navidshaikh commented:
@rhuss  Since we're updating the text printed after service create and update, how about considering https://github.com/knative/client/issues/290 part of this?

rhuss commented:
> @rhuss Since we're updating the text printed after service create and update, how about considering #290 part of this?

Yes, that makes sense for the sync case as the revision would be available at that point, too (we could just add it to the 'showUrl()' method). Or we could add it to the event list, so that the final entry could look like:

```
       12.342s Ready to serve (Revision: myservice-2-3bmz3)
```

Another idea would be to have a `kn service create ... -o yaml` which could output a stream of JSON objects, which then would be even easier to parse in a shell loop (like this kind of [shell-based controller](https://github.com/k8spatterns/examples/blob/ac82b5a875913e24ac08c569ffb38737f7af2321/advanced/Controller/config-watcher-controller.sh#L19-L39).

This would us also allow to not think about how to format human-readable output that it can be easily parsed (which I think is dangerous anyway in the long run as we can easily break things if we make the concrete output format part of a kind of "API")

rhuss commented:
BTW, one reason why I would love to have the URL on a separate line also for human readable output is highlighted in https://youtu.be/9rnrOK0Ifqs?t=1876 : As you can see its not so easy to select an URL in floating text than having a double (or tripple) mouse click on a single line.

Not sure whether we should do that for a revision, too.

Maye we can rephrase the last URL line as:

```
Sevice 'demo' with revision 'demo-1-abced' can be reached at
https://demo.example.com
```

navidshaikh commented:
@rhuss : The URL on a separate (last) line is good, +1 for that.

For,

> Sevice 'demo' with revision 'demo-1-abced' can be reached at

the service might not necessarily be serving from this single revision.

WDYT about this:

For service create:
```
Service 'demo-1' created in namespace 'default'.
Revision 'demo-1-abced' created in namespace 'default'.
[..]

URL:
http://demo.default.example.com
```
For update creating a new revision:
```
Service 'demo-1' updated in namespace 'default'.
Revision 'demo-1-wxyz' created in namespace 'default'.
[..]

URL:
http://demo.default.example.com
```

For update **not** creating a new revision:
```
Service 'demo-1' updated in namespace 'default'.
[..]

URL:
http://demo.default.example.com
```

rhuss commented:
> the service might not necessarily be serving from this single revision.

I thought for `kn service create` we will only have one revision ? (not sure about `kn service create --force though`)

rhuss commented:
The revision can be inferred best only after the service is ready, not when the Service CR is created.

The output for sync create looks like (with a revision added after the service reconciled):

```
Creating service 'svc' in namespace 'default':

    0.123 bla
    1.245 bla
   10.345 Ready to serve.

Revision: 'svc-1-sada'
URL:
https://demo.com
```

I don't think we need to repeat the namespace for the revision name.

rhuss commented:
@navidshaikh I added now the revision name (uncoditionally, but for `update` I could try to find out whether the revision has changed and add an "(unchanged)" if not.

I also updated the asciinema video in the issue comment to reflect this change.

rhuss commented:
@navidshaikh updated now with "(unchanged)" as suffix to a revision if that revision has not changed (also did an update of the video)
